File: fitness_engine/meal_plan/meal_templates.py
"""
Meal templates and macro allocation per meal.

Phase-1 framework: ships with default meal-frequency templates and macro
allocation percentages. The user will supply detailed meal resources later.
"""
from __future__ import annotations

from ..models.meal import MealType

# === Macro allocation per meal (% of daily calories per meal) ===
# Canonical definition lives in STANDARD_ALLOCATIONS in profile_requirements.

# For 5 meals, we add a second snack labeled as SNACK too — we'll handle
# this in the allocator by splitting the snack allocation evenly.
MEAL_ORDER = [
    MealType.BREAKFAST, MealType.SNACK, MealType.LUNCH,
    MealType.SNACK, MealType.DINNER,
]



def get_meal_plan_template(meal_frequency: int) -> list[MealType]:
    """
    Get the ordered list of meal types for a day based on frequency.
    """
    if meal_frequency == 2:
        return [MealType.LUNCH, MealType.DINNER]
    elif meal_frequency == 3:
        return [MealType.BREAKFAST, MealType.LUNCH, MealType.DINNER]
    elif meal_frequency == 4:
        return [MealType.BREAKFAST, MealType.LUNCH, MealType.SNACK, MealType.DINNER]
    elif meal_frequency == 5:
        return [MealType.BREAKFAST, MealType.SNACK, MealType.LUNCH,
                MealType.SNACK, MealType.DINNER]
    else:
        return [MealType.BREAKFAST, MealType.LUNCH, MealType.DINNER]


# === Default meal name templates ===
MEAL_NAMES = {
    MealType.BREAKFAST: {
        "default": "Breakfast",
        "options": ["Greek Yogurt Parfait", "Oatmeal Power Bowl",
                    "Egg & Avocado Toast", "Protein Smoothie",
                    "Veggie Omelette", "Banana & PB Oats",
                    "Cottage Cheese Bowl"],
    },
    MealType.LUNCH: {
        "default": "Lunch",
        "options": ["Grilled Chicken Salad", "Turkey & Rice Bowl",
                    "Tuna Salad Wrap", "Salmon & Quinoa",
                    "Chicken Burrito Bowl", "Steak & Sweet Potato",
                    "Lentil Power Bowl"],
    },
    MealType.DINNER: {
        "default": "Dinner",
        "options": ["Salmon & Asparagus", "Steak & Veggies",
                    "Chicken & Broccoli", "Shrimp Stir-Fry",
                    "Cod & Sweet Potato", "Pork & Green Beans",
                    "Turkey Meatballs & Pasta"],
    },
    MealType.SNACK: {
        "default": "Snack",
        "options": ["Whey Shake & Banana", "Greek Yogurt & Berries",
                    "Apple & Almonds", "Cottage Cheese & Pear",
                    "Protein Bar & Apple", "Edamame Bowl",
                    "Boiled Eggs & Carrots"],
    },
    MealType.PRE_WORKOUT: {
        "default": "Pre-Workout",
        "options": ["Banana & Whey", "Oats & Whey", "Rice Cake & PB"],
    },
    MealType.POST_WORKOUT: {
        "default": "Post-Workout",
        "options": ["Whey & Banana", "Chicken & Rice", "Greek Yogurt & Berries"],
    },
}


def get_meal_name(meal_type: MealType, day: int = 1) -> str:
    """Get a meal name from the template, varying by day for 7-day rotation."""
    options = MEAL_NAMES.get(meal_type, {}).get("options", [meal_type.value])
    return options[(day - 1) % len(options)]


__all__ = [
    "MEAL_ORDER", "MEAL_NAMES",
    "get_meal_plan_template", "get_meal_name",
]

File: download/fitn/fitn_engine/lib/src/meal_plan/meal_templates.dart
/// Meal templates: name cycling per day + meal-type allocation. See §4.5.
library;

import '../models/enums.dart';

/// Meal-type allocation by frequency. See §4.5 step 1.
///
/// 2 meals → LUNCH 0.50, DINNER 0.50 (no breakfast — intentional).
/// 3 meals → BREAKFAST 0.30, LUNCH 0.35, DINNER 0.35.
/// 4 meals → BREAKFAST 0.25, LUNCH 0.30, DINNER 0.30, SNACK 0.15.
/// 5 meals → BREAKFAST 0.20, LUNCH 0.25, DINNER 0.30, SNACK 0.15, SNACK 0.10.
List<MealAllocation> mealAllocationForFrequency(int frequency) {
  switch (frequency) {
    case 2:
      return const [
        MealAllocation(MealType.lunch, 0.50),
        MealAllocation(MealType.dinner, 0.50),
      ];
    case 3:
      return const [
        MealAllocation(MealType.breakfast, 0.30),
        MealAllocation(MealType.lunch, 0.35),
        MealAllocation(MealType.dinner, 0.35),
      ];
    case 4:
      return const [
        MealAllocation(MealType.breakfast, 0.25),
        MealAllocation(MealType.lunch, 0.30),
        MealAllocation(MealType.dinner, 0.30),
        MealAllocation(MealType.snack, 0.15),
      ];
    case 5:
      // Canonical order: BREAKFAST, SNACK, LUNCH, SNACK, DINNER (interleaved).
      return const [
        MealAllocation(MealType.breakfast, 0.20),
        MealAllocation(MealType.snack, 0.15),
        MealAllocation(MealType.lunch, 0.25),
        MealAllocation(MealType.snack, 0.10),
        MealAllocation(MealType.dinner, 0.30),
      ];
    default:
      return const [
        MealAllocation(MealType.breakfast, 0.30),
        MealAllocation(MealType.lunch, 0.35),
        MealAllocation(MealType.dinner, 0.35),
      ];
  }
}

class MealAllocation {
  const MealAllocation(this.mealType, this.fraction);
  final MealType mealType;
  final double fraction;
}

/// Default meal-name templates per meal type. 7 options each, cycled by day.
const Map<MealType, List<String>> defaultMealNames = {
  MealType.breakfast: [
    'Greek Yogurt Parfait',
    'Oatmeal Power Bowl',
    'Egg & Avocado Toast',
    'Protein Smoothie',
    'Cottage Cheese & Berries',
    'Banana Pancakes',
    'Breakfast Burrito',
  ],
  MealType.lunch: [
    'Chicken & Quinoa Bowl',
    'Tofu Stir-Fry',
    'Lentil Soup & Salad',
    'Turkey Wrap',
    'Salmon & Sweet Potato',
    'Beef Stir-Fry',
    'Mediterranean Bowl',
  ],
  MealType.dinner: [
    'Grilled Salmon',
    'Beef Stir-Fry',
    'Alicha Doro Wot',
    'Shiro Wot',
    'Roast Chicken',
    'Tofu Curry',
    'Lentil Stew',
  ],
  MealType.snack: [
    'Cottage Cheese & Berries',
    'Apple & Almonds',
    'Greek Yogurt',
    'Protein Shake',
    'Trail Mix',
    'Hummus & Veggies',
    'Boiled Eggs',
  ],
  MealType.side: [
    'Mixed Green Salad',
    'Roasted Vegetables',
    'Steamed Broccoli',
    'Quinoa Salad',
    'Sweet Potato Mash',
    'Coleslaw',
    'Garlic Bread',
  ],
  MealType.preWorkout: [
    'Banana Oats',
    'Oat Energy Bowl',
    'Apple & Peanut Butter',
    'Rice Cakes',
    'Smoothie',
    'Toast & Honey',
    'Dried Fruit Mix',
  ],
  MealType.postWorkout: [
    'Recovery Smoothie',
    'Greek Yogurt Bowl',
    'Protein Shake',
    'Chicken & Rice',
    'Tofu Bowl',
    'Chocolate Milk',
    'Egg & Avocado Toast',
  ],
};

/// Get the meal name for a (day, mealType) — cycles through 7 default options.
String mealNameFor(int dayNumber, MealType mealType) {
  final options = defaultMealNames[mealType] ?? ['Meal'];
  final idx = (dayNumber - 1) % options.length;
  return options[idx];
}

/// Training-day pattern by training-days-per-week. See §4.5 step 2.
/// 2d → [1, 4], 3d → [1, 3, 5], 4d → [1, 2, 4, 5],
/// 5d → [1, 2, 3, 4, 5], 6d → [1, 2, 3, 4, 5, 6].
List<int> trainingDaysFor(int daysPerWeek) {
  return switch (daysPerWeek) {
    2 => [1, 4],
    3 => [1, 3, 5],
    4 => [1, 2, 4, 5],
    5 => [1, 2, 3, 4, 5],
    6 => [1, 2, 3, 4, 5, 6],
    _ => [1, 3, 5],
  };
}
